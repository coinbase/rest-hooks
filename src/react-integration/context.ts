import React from 'react';
import { ActionTypes } from 'types';
import { initialState } from '~/state/reducer';

export const StateContext = React.createContext(initialState);
export const DispatchContext = React.createContext((value: ActionTypes) => {});
