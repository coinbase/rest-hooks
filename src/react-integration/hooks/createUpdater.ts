import { OptimisticUpdateFunction } from '~/types';

type SimpleUpdateFunction = (fromResults: any, toResults: any) => any;

/**
 * Function generator that creates a merge function that merges an entity into results
 *
 * @param fromSchema The schema for the entity that we want to add to the toSchema
 * @param toSchema The schema that represents the results we are merging into
 * @param updateFunction The user-provided update function that merges the entity
 *   returned in the fromSchema into the results of shape toSchema
 */
export default function createUpdater(
  fromSchema: any,
  toSchema: any,
  updateFunction: SimpleUpdateFunction,
): OptimisticUpdateFunction {
  console.log(fromSchema, toSchema, updateFunction);
  return (fromResults: any, resultsFromList: any): any => {
    // TODO: Correctly match entities
    return resultsFromList;
  };
}
