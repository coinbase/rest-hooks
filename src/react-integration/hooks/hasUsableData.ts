import { FetchShape, Schema } from '~/resource';

/** If the invalidIfStale option is set we suspend if resource has expired */
export default function hasUsableData<
  S extends Schema,
  Params extends Readonly<object>,
  Body extends Readonly<object | string> | void
>(
  cacheReady: boolean,
  fetchShape: Pick<FetchShape<S, Params, Body>, 'options'>,
) {
  return !(
    (fetchShape.options && fetchShape.options.invalidIfStale) ||
    !cacheReady
  );
}
