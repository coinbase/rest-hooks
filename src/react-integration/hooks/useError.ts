import { ReadShape, Schema, RequestResource } from '~/resource';
import useMeta from './useMeta';

/** Access a resource or error if failed to get it */
export default function useError<
  Params extends Readonly<object>,
  Body extends Readonly<object | string> | void,
  S extends Schema
>(
  fetchShape: ReadShape<S, Params, Body>,
  params: Params | null,
  resource: RequestResource<typeof fetchShape> | null,
) {
  const meta = useMeta(fetchShape, params);
  if (!resource) {
    if (!meta) return;
    if (!meta.error) {
      // this means we probably deleted the entity found in this result
      const err: any = new Error(
        `Resource not found ${params ? fetchShape.getFetchKey(params) : ''}`,
      );
      err.status = 404;
      return err;
    } else {
      return meta.error;
    }
  }
}
