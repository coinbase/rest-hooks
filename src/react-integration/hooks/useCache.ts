import { useContext } from 'react';

import { StateContext } from '~/react-integration/context';
import { ReadShape, Schema } from '~/resource';
import { useSchemaSelect } from '~/state/selectors';

/** Access a resource if it is available. */
export default function useCache<
  Params extends Readonly<object>,
  S extends Schema
>(
  fetchShape: Pick<ReadShape<S, Params, any>, 'schema' | 'getFetchKey'>,
  params: Params | null,
) {
  const state = useContext(StateContext);
  return useSchemaSelect(fetchShape, state, params);
}
