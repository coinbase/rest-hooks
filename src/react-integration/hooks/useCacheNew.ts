import { useContext } from 'react';

import { StateContext } from '~/react-integration/context';
import { ReadShape, Schema } from '~/resource';
import { useDenormalized } from '~/state/selectors';

/** Access a resource if it is available. */
export default function useCacheNew<
  Params extends Readonly<object>,
  S extends Schema
>(
  fetchShape: Pick<ReadShape<S, Params>, 'schema' | 'getFetchKey'>,
  params: Params | null,
) {
  const state = useContext(StateContext);
  return useDenormalized(fetchShape, params, state)[0];
}
