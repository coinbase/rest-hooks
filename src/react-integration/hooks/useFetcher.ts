import { useContext, useCallback } from 'react';

import { RequestShape, Schema, isDeleteShape } from '../../resource';
import { DispatchContext } from '../context';

const SHAPE_TYPE_TO_RESPONSE_TYPE: Record<
  RequestShape<any, any, any>['type'],
  'receive' | 'rpc' | 'purge'
> = {
  read: 'receive',
  mutate: 'rpc',
  delete: 'purge',
};

/** Build an imperative dispatcher to issue network requests. */
export default function useFetcher<
  Params extends Readonly<object>,
  Body extends Readonly<object> | void,
  S extends Schema
>(requestShape: RequestShape<S, Params, Body>, throttle = false) {
  const { fetch, schema, type, getUrl, options } = requestShape;
  const responseType = SHAPE_TYPE_TO_RESPONSE_TYPE[type];

  const dispatch = useContext(DispatchContext);

  const fetchDispatcher = useCallback(
    (body: Body, params: Params) => {
      const url = getUrl(params);
      const identifier = isDeleteShape(requestShape)
        ? (schema as any).getId(params)
        : url;
      let resolve: (value?: any | PromiseLike<any>) => void = () => undefined;
      let reject: (reason?: any) => void = () => undefined;
      const promise = new Promise<any>((a, b) => {
        [resolve, reject] = [a, b];
      });

      dispatch({
        type: 'fetch',
        payload: () => fetch(url, body),
        meta: {
          schema,
          responseType,
          url: identifier,
          throttle,
          options,
          resolve,
          reject,
        },
      });
      return promise;
    },
    [fetch, schema, type, getUrl, responseType, throttle, dispatch],
  );
  return fetchDispatcher;
}
