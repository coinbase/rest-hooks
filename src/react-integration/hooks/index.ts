import useFetcher from './useFetcher';
import useCache from './useCache';
import useRetrieve from './useRetrieve';
import useResource from './useResource';
import useSubscription from './useSubscription';
import useMeta from './useMeta';
import useError from './useError';
import useInvalidator from './useInvalidator';
import useResetter from './useResetter';

export {
  useFetcher,
  useCache,
  useError,
  useRetrieve,
  useResource,
  useSubscription,
  useMeta,
  useInvalidator,
  useResetter,
};
