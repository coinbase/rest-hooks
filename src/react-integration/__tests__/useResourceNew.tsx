import React, { Suspense } from 'react';
import { render } from '@testing-library/react';
import nock from 'nock';
import { normalize } from '../../resource';

import { DispatchContext, StateContext } from '../context';
import {
  CoolerArticleResource,
  UserResource,
  InvalidIfStaleArticleResource,
} from '../../__tests__/common';
import { useResourceNew } from '../hooks';
import makeRenderRestHook from '../../test/makeRenderRestHook';
import { makeCacheProvider } from '../../test/providers';
import mockInitialState from '../../test/mockState';
import { payload, users, nested } from './fixtures';

async function testDispatchFetch(
  Component: React.FunctionComponent<any>,
  payloads: any[],
) {
  const dispatch = jest.fn();
  const tree = (
    <DispatchContext.Provider value={dispatch}>
      <Suspense fallback={null}>
        <Component />
      </Suspense>
    </DispatchContext.Provider>
  );
  render(tree);
  expect(dispatch).toHaveBeenCalled();
  expect(dispatch.mock.calls.length).toBe(payloads.length);
  let i = 0;
  for (const call of dispatch.mock.calls) {
    expect(call[0]).toMatchSnapshot();
    const action = call[0];
    const res = await action.payload();
    expect(res).toEqual(payloads[i]);
    i++;
  }
}

function ArticleComponentTester({ invalidIfStale = false }) {
  const resource = invalidIfStale
    ? InvalidIfStaleArticleResource
    : CoolerArticleResource;
  const article = useResourceNew(resource.detailShape(), {
    id: payload.id,
  });
  return (
    <div>
      <h3>{article.title}</h3>
      <p>{article.content}</p>
    </div>
  );
}

describe('useResourceNew()', () => {
  let renderRestHook: ReturnType<typeof makeRenderRestHook>;
  const fbmock = jest.fn();

  function Fallback() {
    fbmock();
    return null;
  }

  beforeEach(() => {
    renderRestHook = makeRenderRestHook(makeCacheProvider);
    nock('http://test.com')
      .get(`/article-cooler/${payload.id}`)
      .reply(200, payload);
    nock('http://test.com')
      .delete(`/article-cooler/${payload.id}`)
      .reply(204, '');
    nock('http://test.com')
      .delete(`/article/${payload.id}`)
      .reply(200, {});
    nock('http://test.com')
      .get(`/article-cooler/0`)
      .reply(403, {});
    nock('http://test.com')
      .get(`/article-cooler/666`)
      .reply(200, '');
    nock('http://test.com')
      .get(`/article-cooler/`)
      .reply(200, nested);
    nock('http://test.com')
      .get(`/user/`)
      .reply(200, users);
  });
  afterEach(() => {
    nock.cleanAll();
    renderRestHook.cleanup();
  });

  it('should dispatch an action that fetches', async () => {
    await testDispatchFetch(ArticleComponentTester, [payload]);
  });

  it('should dispatch fetch when sent multiple arguments', async () => {
    function MultiResourceTester() {
      const [article, user] = useResourceNew(
        [
          CoolerArticleResource.detailShape(),
          {
            id: payload.id,
          },
        ],
        [UserResource.listShape(), {}],
      );
      return null;
    }
    await testDispatchFetch(MultiResourceTester, [payload, users]);
  });
  it('should throw same promise until both resolve', async () => {
    const renderRestHook = makeRenderRestHook(makeCacheProvider);
    jest.useFakeTimers();
    nock('http://test.com')
      .get(`/article-cooler/${payload.id}`)
      .delay(1000)
      .reply(200, payload);
    nock('http://test.com')
      .get(`/user/`)
      .delay(2000)
      .reply(200, users);

    function MultiResourceTester() {
      try {
        const [article, user] = useResourceNew(
          [
            CoolerArticleResource.detailShape(),
            {
              id: payload.id,
            },
          ],
          [UserResource.listShape(), {}],
        );
        return article;
      } catch (e) {
        // TODO: we're not handling suspense properly so react complains
        // When upgrading test util we should be able to fix this as we'll suspense ourselves.
        if (typeof e.then === 'function') {
          return e;
        } else {
          throw e;
        }
      }
    }
    const { rerender, result } = renderRestHook(MultiResourceTester);
    const firstPromise = result.current;
    expect(firstPromise).toBeDefined();
    expect(typeof firstPromise.then).toBe('function');

    jest.advanceTimersByTime(50);
    rerender();
    expect(result.current).toBe(firstPromise);
    jest.advanceTimersByTime(1000);
    rerender();
    expect(result.current).toBe(firstPromise);
    jest.advanceTimersByTime(2000);
    rerender();
    expect(result.current).toBe(firstPromise);

    // TODO: we're not handling suspense properly so react complains
    // When upgrading test util we should be able to fix this as we'll suspense ourselves.
    const oldError = console.error;
    console.error = () => {};
    jest.runAllTimers();
    await result.current;
    rerender();
    expect(result.current).toMatchInlineSnapshot(`
      CoolerArticleResource {
        "author": null,
        "content": "whatever",
        "id": 5,
        "tags": Array [
          "a",
          "best",
          "react",
        ],
        "title": "hi ho",
      }
    `);
    // eslint-disable-next-line require-atomic-updates
    console.error = oldError;
  });
  it('should NOT suspend if result already in cache and options.invalidIfStale is false', () => {
    const state = mockInitialState([
      {
        request: CoolerArticleResource.detailShape(),
        params: payload,
        result: payload,
      },
    ]);

    const tree = (
      <StateContext.Provider value={state}>
        <Suspense fallback={<Fallback />}>
          <ArticleComponentTester />
        </Suspense>
      </StateContext.Provider>
    );
    const { getByText } = render(tree);
    expect(fbmock).not.toBeCalled();
    const title = getByText(payload.title);
    expect(title).toBeDefined();
    expect(title.tagName).toBe('H3');
  });
  it('should NOT suspend even when result is stale and options.invalidIfStale is false', () => {
    const { entities, result } = normalize(
      payload,
      CoolerArticleResource.getEntitySchema(),
    );
    const fetchKey = CoolerArticleResource.detailShape().getFetchKey(payload);
    const state = {
      entities,
      results: {
        [fetchKey]: result,
      },
      meta: {
        [fetchKey]: {
          date: 0,
          expiresAt: 0,
        },
      },
    };

    const tree = (
      <StateContext.Provider value={state}>
        <DispatchContext.Provider value={() => {}}>
          <Suspense fallback={<Fallback />}>
            <ArticleComponentTester />
          </Suspense>{' '}
        </DispatchContext.Provider>
      </StateContext.Provider>
    );
    const { getByText } = render(tree);
    expect(fbmock).not.toBeCalled();
    const title = getByText(payload.title);
    expect(title).toBeDefined();
    expect(title.tagName).toBe('H3');
  });
  it('should NOT suspend if result is not stale and options.invalidIfStale is true', () => {
    const { entities, result } = normalize(
      payload,
      InvalidIfStaleArticleResource.getEntitySchema(),
    );
    const fetchKey = InvalidIfStaleArticleResource.detailShape().getFetchKey(
      payload,
    );
    const state = {
      entities,
      results: {
        [fetchKey]: result,
      },
      meta: {
        [fetchKey]: {
          date: Infinity,
          expiresAt: Infinity,
        },
      },
    };

    const tree = (
      <StateContext.Provider value={state}>
        <Suspense fallback={<Fallback />}>
          <ArticleComponentTester invalidIfStale />
        </Suspense>
      </StateContext.Provider>
    );
    const { getByText } = render(tree);
    expect(fbmock).not.toBeCalled();
    const title = getByText(payload.title);
    expect(title).toBeDefined();
    expect(title.tagName).toBe('H3');
  });
  it('should suspend if result stale in cache and options.invalidIfStale is true', () => {
    const { entities, result } = normalize(
      payload,
      InvalidIfStaleArticleResource.getEntitySchema(),
    );
    const fetchKey = InvalidIfStaleArticleResource.detailShape().getFetchKey(
      payload,
    );
    const state = {
      entities,
      results: {
        [fetchKey]: result,
      },
      meta: {
        [fetchKey]: {
          date: 0,
          expiresAt: 0,
        },
      },
    };

    const tree = (
      <StateContext.Provider value={state}>
        <DispatchContext.Provider value={() => {}}>
          <Suspense fallback={<Fallback />}>
            <ArticleComponentTester invalidIfStale />
          </Suspense>
        </DispatchContext.Provider>
      </StateContext.Provider>
    );
    render(tree);
    expect(fbmock).toHaveBeenCalled();
  });

  // taken from integration
  it('useResourceNew() should throw errors on bad network', async () => {
    const { result, waitForNextUpdate } = renderRestHook(() => {
      return useResourceNew(CoolerArticleResource.detailShape(), {
        title: '0',
      });
    });
    expect(result.current).toBe(null);
    await waitForNextUpdate();
    expect(result.error).toBeDefined();
    expect((result.error as any).status).toBe(403);
  });

  it('useResourceNew() should throw errors on bad network (multiarg)', async () => {
    const { result, waitForNextUpdate } = renderRestHook(() => {
      return useResourceNew([
        CoolerArticleResource.detailShape(),
        {
          title: '0',
        },
      ]);
    });
    expect(result.current).toBe(null);
    await waitForNextUpdate();
    expect(result.error).toBeDefined();
    expect((result.error as any).status).toBe(403);
  });

  it('should resolve parallel useResourceNew() request', async () => {
    const { result, waitForNextUpdate } = renderRestHook(() => {
      return useResourceNew(
        [
          CoolerArticleResource.detailShape(),
          {
            id: payload.id,
          },
        ],
        [UserResource.listShape(), {}],
      );
    });
    await waitForNextUpdate();
    const [article, users] = result.current;
    expect(article instanceof CoolerArticleResource).toBe(true);
    expect(article.title).toBe(payload.title);
    expect(users).toBeDefined();
    expect(users.length).toBeTruthy();
    expect(users[0] instanceof UserResource).toBe(true);
  });

  it('should not suspend with no params to useResourceNew()', () => {
    let article: any;
    const { result } = renderRestHook(() => {
      article = useResourceNew(CoolerArticleResource.detailShape(), null);
      return 'done';
    });
    expect(result.current).toBe('done');
    expect(article).toBeUndefined();
  });
});
