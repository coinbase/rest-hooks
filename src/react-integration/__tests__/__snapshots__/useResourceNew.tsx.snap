// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useResourceNew() should dispatch an action that fetches 1`] = `
Object {
  "meta": Object {
    "options": undefined,
    "reject": [Function],
    "resolve": [Function],
    "responseType": "rest-hooks/receive",
    "schema": EntitySchema {
      "_getId": [Function],
      "_idAttribute": [Function],
      "_key": "http://test.com/article-cooler/",
      "_mergeStrategy": [Function],
      "_processStrategy": [Function],
      "denormalize": [Function],
      "schema": Object {},
    },
    "throttle": true,
    "url": "GET http://test.com/article-cooler/5",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useResourceNew() should dispatch fetch when sent multiple arguments 1`] = `
Object {
  "meta": Object {
    "options": undefined,
    "reject": [Function],
    "resolve": [Function],
    "responseType": "rest-hooks/receive",
    "schema": EntitySchema {
      "_getId": [Function],
      "_idAttribute": [Function],
      "_key": "http://test.com/article-cooler/",
      "_mergeStrategy": [Function],
      "_processStrategy": [Function],
      "denormalize": [Function],
      "schema": Object {},
    },
    "throttle": true,
    "url": "GET http://test.com/article-cooler/5",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useResourceNew() should dispatch fetch when sent multiple arguments 2`] = `
Object {
  "meta": Object {
    "options": undefined,
    "reject": [Function],
    "resolve": [Function],
    "responseType": "rest-hooks/receive",
    "schema": Array [
      EntitySchema {
        "_getId": [Function],
        "_idAttribute": [Function],
        "_key": "http://test.com/user/",
        "_mergeStrategy": [Function],
        "_processStrategy": [Function],
        "denormalize": [Function],
        "schema": Object {},
      },
    ],
    "throttle": true,
    "url": "GET http://test.com/user/",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useResourceNewNew() should dispatch an action that fetches 1`] = `
Object {
  "meta": Object {
    "options": undefined,
    "reject": [Function],
    "resolve": [Function],
    "responseType": "rest-hooks/receive",
    "schema": EntitySchema {
      "_getId": [Function],
      "_idAttribute": [Function],
      "_key": "http://test.com/article-cooler/",
      "_mergeStrategy": [Function],
      "_processStrategy": [Function],
      "denormalize": [Function],
      "schema": Object {},
    },
    "throttle": true,
    "url": "GET http://test.com/article-cooler/5",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useResourceNewNew() should dispatch fetch when sent multiple arguments 1`] = `
Object {
  "meta": Object {
    "options": undefined,
    "reject": [Function],
    "resolve": [Function],
    "responseType": "rest-hooks/receive",
    "schema": EntitySchema {
      "_getId": [Function],
      "_idAttribute": [Function],
      "_key": "http://test.com/article-cooler/",
      "_mergeStrategy": [Function],
      "_processStrategy": [Function],
      "denormalize": [Function],
      "schema": Object {},
    },
    "throttle": true,
    "url": "GET http://test.com/article-cooler/5",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useResourceNewNew() should dispatch fetch when sent multiple arguments 2`] = `
Object {
  "meta": Object {
    "options": undefined,
    "reject": [Function],
    "resolve": [Function],
    "responseType": "rest-hooks/receive",
    "schema": Array [
      EntitySchema {
        "_getId": [Function],
        "_idAttribute": [Function],
        "_key": "http://test.com/user/",
        "_mergeStrategy": [Function],
        "_processStrategy": [Function],
        "denormalize": [Function],
        "schema": Object {},
      },
    ],
    "throttle": true,
    "url": "GET http://test.com/user/",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;
